README.md

```diff
@@ -18,15 +18,49 @@ Once you have implemented the `sa` program (and tested it to the best of your ab
+
+The half implementation is an attempt at skew, that we didnt manage to finish in time, but you can have a look and see if it tickles your fancy, viewer discretion is adviced. 
 
 ## Insights you may have had while implementing the algorithm
+When you finish up writing a bucket sort for the third time in a day, you begin to realize that maybe you should have just made a function for it to begin with, fool me once... 
+
+In general, we should get better at recognizing when supproblems can be boiled down to a standalone function to make the code more readable, turns out having somewhat readable code also helps with debugging, and being anle to trace errors back to specific functions also speeds up that process
+
 
 ## Problems encountered if any
+The usual problem with Johan not being able to sync with git, implementing the skew algorithm turned out to be tricky just in general. The biggest problems with skew being, understanding what to do with the results of the recursion, and mapping back and forth correctly. 
+There is ofcourse the issue of a the horribly slow radix implementation, but the only ones who suffer from that would be us, when we have to run test, and any soul unfortunate enough to take a look at the code.
 
 ## Validation
 
 *How did you validate that the construct and the search algorithm works?*
 
+We once again used Johan's beautiful data generators to generate various strings including repetetive, empty and identical and palindromic.
+Compared with the output of the naive pattern matching we implemented in the first project.
+We should probably start using the pytest to automate this.
+
 ## Running time
 
 *List experiments and results that show that both the construction algorithm and the search algorithm works in the expected running time. Add figures by embedding them here, as you learned how to do in project 1.*
+
+
+The running time seems to behave as expected. The radix sort implementation of array construction should be O(n^2), and it is when run on data generated from a simple markov model using data_gen.py.
+
+![](graphs/array_construction_not_linearized.png)
+
+After dividing running times by $n^2$ for each datapoint we get a linear graph increasing with what is assumed to be the number of outputs (z)
+
+![](graphs/array_construction_linearized.png)
+
+For the pattern matching we did a similar experiment using the exact same data and a random string of length 5 as the search pattern. It is expected to run in $O(m*log(n)+z)$. Since we are keeping m constant (5), the curve should be logarithmic. 
+
+![](graphs/pattern_match_not_linearized.png)
+
+To test if the curve is indeed logarithmic the times are divided by $log(n)$.
+
+![](graphs/pattern_match_linarized.png)
+
+There is a curve to this graph which we cannot explain. It might have something to do with the outputs (z).
```

Comment:
Again, this is a lame excuse. You can easily leave out the $z$ part of the search. Just spend the $O(m\log n)$ time to find the interval to report, then turn off the clock while outputting (or skip outputting entirely if you want).

It _could_ be true, and it probably is, but it is very easy to verify, and then we don't have to guess.